plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,146 @@
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStateChangedObservers = CopyOnWriteArrayList<ViewportStateChangedObserver>()
+
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+
+  override val states : Set<ViewportState>
+    get() = internalStates.toSet()
+
+  override var currentState: ViewportState? = null
+    private set(value) {
+      field?.cleanUp()
+      field = value
+      notifyStateChanged()
+    }
+
+  private fun notifyStateChanged() {
+    registeredStateChangedObservers.forEach { it.onViewportStateChanged(currentState) }
+  }
+
+  override fun addViewportState(state: ViewportState) {
+    internalStates.add(state)
+  }
+
+  override fun removeViewportState(state: ViewportState) {
+    if (currentState === state) {
+      currentState = null
+    }
+    internalStates.remove(state)
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      state,
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_STATE_TRANSITION_MAX_DURATION_MS)
+        .apply(stateTransitionOptionsBlock).build(),
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+        .apply(frameTransitionOptionsBlock).build(),
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    internalStates.add(state)
+    state.transitionToState(
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+    currentState = state
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptionsBlock,
+      frameTransitionOptionsBlock,
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToIdle() {
+    currentState = null
+  }
```

Comment:
Why not make IDLE its own type? That would allow us to run joint `when` clauses and other type operations while making the `currentState` non-null.

Reply:
Making the IDLE its own type would complicate current design, as we ask users to create ViewportState instances before usage. And there wouldn't be any transition to IDLE state, and it's also not an async operation, where most of our APIs are designed for.

We could debate these out, but @andrew and I had consensus that using a dedicated IDLE state might not bring enough value.